File: software/ekf/ekf_real/ekf.cpp
#include "ekf.h"

static void ekf_g(ekf_t *ekf, ekf_control_t *c);
static void ekf_dg_dstate(ekf_t *ekf, ekf_control_t *c);
static void ekf_dg_dcontrol(ekf_t *ekf, ekf_control_t *c);

static void ekf_dh_dstate(ekf_t *ekf, ekf_measurement_t *m);
static void ekf_h(ekf_t *ekf, ekf_measurement_t *m);

void ekf_init(ekf_t *ekf, ekf_params_t *params){
    ekf->params = params;

    ekf->G[0][0] = 1;
    ekf->G[0][1] = 0;
    ekf->G[0][2] = 0;
    ekf->G[1][0] = 0;
    ekf->G[1][1] = 1;
    ekf->G[1][2] = 0;
    ekf->G[2][0] = 0;
    ekf->G[2][1] = 0;
    ekf->G[2][2] = 1;

    if(params->tag_cart_x == 0){
        params->tag_polar_r = params->tag_cart_y;
        params->tag_polar_theta = params->tag_cart_y > 0 ? M_PI/2 : -M_PI/2;
    }else{
        params->tag_polar_r = atan2(params->tag_cart_y, params->tag_cart_x);
        params->tag_polar_theta = sqrt(params->tag_cart_x*params->tag_cart_x + params->tag_cart_y*params->tag_cart_y);
    }

    ekf->Q[0][0] = params->measurement_distance_stddev * params->measurement_distance_stddev;
    ekf->Q[1][1] = params->measurement_angle_stddev * params->measurement_angle_stddev;
    ekf->Q[0][1] = ekf->Q[1][0] = 0;
}

void ekf_predict(ekf_t *ekf, ekf_control_t *c){
    ekf_g(ekf,c);

    ekf_dg_dstate(ekf,c);
    ekf_dg_dcontrol(ekf, c);

    float gsgt[3][3];
    gsgt[0][0] = ekf->cov[0][0] + 2*ekf->G[0][2]*ekf->cov[0][2] + ekf->G[0][2]*ekf->G[0][2]*ekf->cov[2][2];
    gsgt[1][1] = ekf->cov[1][1] + 2*ekf->G[1][2]*ekf->cov[1][2] + ekf->G[1][2]*ekf->G[1][2]*ekf->cov[2][2];
    gsgt[2][2] = ekf->cov[2][2];
    gsgt[0][1] = gsgt[1][0] = ekf->cov[0][1] + ekf->G[0][2]*ekf->cov[1][2] + ekf->G[1][2]*ekf->cov[0][2] + ekf->G[0][2]*ekf->G[1][2]*ekf->cov[2][2];
    gsgt[0][2] = gsgt[2][0] = ekf->cov[0][2] + ekf->G[0][2]*ekf->cov[2][2];
    gsgt[1][2] = gsgt[2][1] = ekf->cov[1][2] + ekf->G[1][2]*ekf->cov[2][2];

    float ll = ekf->params->control_motion_factor*c->l;
    float rr = ekf->params->control_motion_factor*c->r;
    float lr = ekf->params->control_turn_factor*(c->l-c->r);
    float A = ll*ll + lr*lr;
    float B = rr*rr + lr*lr;

    float vsvt[3][3];
    vsvt[0][0] = ekf->V[0][0]*ekf->V[0][0]*A*A + ekf->V[0][1]*ekf->V[0][1]*B*B;
    vsvt[1][1] = ekf->V[1][0]*ekf->V[1][0]*A*A + ekf->V[1][1]*ekf->V[1][1]*B*B;
    vsvt[2][2] = ekf->V[2][0]*ekf->V[2][0]*A*A + ekf->V[2][1]*ekf->V[2][1]*B*B;
    vsvt[0][1] = vsvt[1][0] = ekf->V[0][0]*ekf->V[1][0]*A*A + ekf->V[0][1]*ekf->V[1][1]*B*B;
    vsvt[0][2] = vsvt[2][0] = ekf->V[0][0]*ekf->V[2][0]*A*A + ekf->V[0][1]*ekf->V[2][1]*B*B;
    vsvt[1][2] = vsvt[2][1] = ekf->V[1][0]*ekf->V[2][0]*A*A + ekf->V[2][1]*ekf->V[1][1]*B*B;

    ekf->cov[0][0] = gsgt[0][0] + vsvt[0][0];
    ekf->cov[1][1] = gsgt[1][1] + vsvt[1][1];
    ekf->cov[2][2] = gsgt[2][2] + vsvt[2][2];
    ekf->cov[0][1] = ekf->cov[1][0] = gsgt[0][1] + vsvt[0][1];
    ekf->cov[0][2] = ekf->cov[2][0] = gsgt[0][2] + vsvt[0][2];
    ekf->cov[1][2] = ekf->cov[2][1] = gsgt[1][2] + vsvt[1][2];
}

#include <stdio.h>

void ekf_correct(ekf_t *ekf, ekf_measurement_t *m){

    ekf_dh_dstate(ekf, m);

    float sigma_ht[3][2];
    sigma_ht[0][0] = ekf->cov[0][0]*ekf->H[0][0] + ekf->cov[0][1]*ekf->H[0][1] + ekf->cov[0][2]*ekf->H[0][2];
    sigma_ht[0][1] = ekf->cov[0][0]*ekf->H[1][0] + ekf->cov[0][1]*ekf->H[1][1] + ekf->cov[0][2]*ekf->H[1][2];
    sigma_ht[1][0] = ekf->cov[1][0]*ekf->H[0][0] + ekf->cov[1][1]*ekf->H[0][1] + ekf->cov[1][2]*ekf->H[0][2];
    sigma_ht[1][1] = ekf->cov[1][0]*ekf->H[1][0] + ekf->cov[1][1]*ekf->H[1][1] + ekf->cov[1][2]*ekf->H[1][2];
    sigma_ht[2][0] = ekf->cov[2][0]*ekf->H[0][0] + ekf->cov[2][1]*ekf->H[0][1] + ekf->cov[2][2]*ekf->H[0][2];
    sigma_ht[2][1] = ekf->cov[2][0]*ekf->H[1][0] + ekf->cov[2][1]*ekf->H[1][1] + ekf->cov[2][2]*ekf->H[1][2];

    float h_sigma_ht[2][2];
    h_sigma_ht[0][0] = ekf->H[0][0]*sigma_ht[0][0] + ekf->H[0][1]*sigma_ht[1][0] + ekf->H[0][2]*sigma_ht[2][0];
    h_sigma_ht[1][1] = ekf->H[1][0]*sigma_ht[0][1] + ekf->H[1][1]*sigma_ht[1][1] + ekf->H[1][2]*sigma_ht[2][1];

    h_sigma_ht[0][1] = ekf->H[0][0]*sigma_ht[0][1] + ekf->H[0][1]*sigma_ht[1][1] + ekf->H[0][2]*sigma_ht[2][1];
    h_sigma_ht[1][0] = ekf->H[1][0]*sigma_ht[0][0] + ekf->H[1][1]*sigma_ht[1][0] + ekf->H[1][2]*sigma_ht[2][0];

    float D[2][2];
    D[0][0] = h_sigma_ht[0][0] + ekf->Q[0][0];
    D[0][1] = h_sigma_ht[0][1] + ekf->Q[0][1];
    D[1][0] = h_sigma_ht[1][0] + ekf->Q[1][0];
    D[1][1] = h_sigma_ht[1][1] + ekf->Q[1][1];
    printf("D: %f %f %f %f\r\n", D[0][0], D[0][1], D[1][0], D[1][1]);

    float detD_inv = 1.0f/(D[0][0]*D[1][1] - D[1][0]*D[1][1]);
    printf("detD_inv: %f\r\n", detD_inv);
    
    float D_inv[2][2];
    D_inv[0][0] = D[1][1]*detD_inv;
    D_inv[1][1] = D[0][0]*detD_inv;
    D_inv[0][1] = -D[0][1]*detD_inv;
    D_inv[1][0] = -D[1][0]*detD_inv;
    printf("D: %f %f %f %f\r\n", D_inv[0][0], D_inv[0][1], D_inv[1][0], D_inv[1][1]);

    float K[3][2];
    K[0][0] = sigma_ht[0][0]*D_inv[0][0] + sigma_ht[0][1]*D_inv[1][0];
    K[0][1] = sigma_ht[0][0]*D_inv[0][1] + sigma_ht[0][1]*D_inv[1][1];
    K[1][0] = sigma_ht[1][0]*D_inv[0][0] + sigma_ht[1][1]*D_inv[1][0];
    K[1][1] = sigma_ht[1][0]*D_inv[0][1] + sigma_ht[1][1]*D_inv[1][1];
    K[2][0] = sigma_ht[2][0]*D_inv[0][0] + sigma_ht[2][1]*D_inv[1][0];
    K[2][1] = sigma_ht[2][0]*D_inv[0][1] + sigma_ht[2][1]*D_inv[1][1];
    printf("K: %f %f %f %f %f %f\r\n", K[0][0], K[0][1], K[1][0], K[1][1], K[2][0], K[2][1]);

    ekf_h(ekf, m);
    float innovation[2];
    innovation[0] = m->range - ekf->h[0];
    innovation[1] = m->yaw - ekf->h[1];
    printf("yaw: %f %f\r\n", m->yaw, ekf->h[1]);
    innovation[1] = fmod(innovation[1] + M_PI, 2*M_PI) - M_PI;
    printf("innovation: %f %f\r\n", innovation[0], innovation[1]);

    ekf->x += K[0][0]*innovation[0] + K[0][1]*innovation[1];
    ekf->y += K[1][0]*innovation[0] + K[1][1]*innovation[1];
    ekf->theta += K[2][0]*innovation[0] + K[2][1]*innovation[1];

    float T[3][3];
    T[0][0] = K[0][0]*ekf->H[0][0] + K[0][1]*ekf->H[1][0];
    T[0][1] = K[0][0]*ekf->H[0][1] + K[0][1]*ekf->H[1][1];
    T[0][2] = K[0][0]*ekf->H[0][2] + K[0][1]*ekf->H[1][2];
    T[1][0] = K[1][0]*ekf->H[0][0] + K[1][1]*ekf->H[1][0];
    T[1][1] = K[1][0]*ekf->H[0][1] + K[1][1]*ekf->H[1][1];
    T[1][2] = K[1][0]*ekf->H[0][2] + K[1][1]*ekf->H[1][2];
    T[2][0] = K[2][0]*ekf->H[0][0] + K[2][1]*ekf->H[1][0];
    T[2][1] = K[2][0]*ekf->H[0][1] + K[2][1]*ekf->H[1][1];
    T[2][2] = K[2][0]*ekf->H[0][2] + K[2][1]*ekf->H[1][2];

    T[0][0] = 1 - T[0][0];
    T[1][1] = 1 - T[1][1]; 
    T[2][2] = 1 - T[2][2];
    T[0][1] = -T[0][1];
    T[0][2] = -T[0][2];
    T[1][0] = -T[1][0];
    T[1][2] = -T[1][2];
    T[2][0] = -T[2][0];
    T[2][1] = -T[2][1];

    ekf->cov[0][0] = T[0][0]*ekf->cov[0][0] + T[0][1]*ekf->cov[1][0] + T[0][2]*ekf->cov[2][0];
    ekf->cov[0][1] = T[0][0]*ekf->cov[0][1] + T[0][1]*ekf->cov[1][1] + T[0][2]*ekf->cov[2][1];
    ekf->cov[0][2] = T[0][0]*ekf->cov[0][2] + T[0][1]*ekf->cov[1][2] + T[0][2]*ekf->cov[2][2];
    ekf->cov[1][0] = T[1][0]*ekf->cov[0][0] + T[1][1]*ekf->cov[1][0] + T[1][2]*ekf->cov[2][0];
    ekf->cov[1][1] = T[1][0]*ekf->cov[0][1] + T[1][1]*ekf->cov[1][1] + T[1][2]*ekf->cov[2][1];
    ekf->cov[1][2] = T[1][0]*ekf->cov[0][2] + T[1][1]*ekf->cov[1][2] + T[1][2]*ekf->cov[2][2];
    ekf->cov[2][0] = T[2][0]*ekf->cov[0][0] + T[2][1]*ekf->cov[1][0] + T[2][2]*ekf->cov[2][0];
    ekf->cov[2][1] = T[2][0]*ekf->cov[0][1] + T[2][1]*ekf->cov[1][1] + T[2][2]*ekf->cov[2][1];
    ekf->cov[2][2] = T[2][0]*ekf->cov[0][2] + T[2][1]*ekf->cov[1][2] + T[2][2]*ekf->cov[2][2];
}

void ekf_g(ekf_t *ekf, ekf_control_t *c){
    if (c->r != c->l){
        float alpha = (c->r - c->l) / ekf->params->robot_width;
        float rad = c->l/alpha;
        ekf->x += (rad + ekf->params->robot_width/2.0f)*(sin(ekf->theta+alpha) - sin(ekf->theta));
        ekf->y += (rad + ekf->params->robot_width/2.0f)*(-cos(ekf->theta+alpha) + cos(ekf->theta));
        ekf->theta = fmod(ekf->theta + alpha + M_PI, 2*M_PI) - M_PI;
    }
    else{
        ekf->x += c->l * cos(ekf->theta);
        ekf->y += c->l * sin(ekf->theta);
    }
}

void ekf_dg_dstate(ekf_t *ekf, ekf_control_t *c){
    if (c->r != c->l){
        float alpha = (c->r-c->l)/ekf->params->robot_width;
        float theta_ = ekf->theta + alpha;
        float rpw2 = c->l/alpha + ekf->params->robot_width/2.0;
        
        ekf->G[0][2] = rpw2*(cos(theta_) - cos(ekf->theta));
        ekf->G[1][2] = rpw2*(sin(theta_) - sin(ekf->theta));
    }
    else{
        ekf->G[0][2] = -c->l*sin(ekf->theta);
        ekf->G[1][2] =  c->l*cos(ekf->theta);
    }
    printf("G: %f %f %f %f %f %f\r\n", ekf->G[0][0], ekf->G[0][1], ekf->G[1][0], ekf->G[1][1], ekf->G[2][0], ekf->G[2][1]);
}

void ekf_dg_dcontrol(ekf_t *ekf, ekf_control_t *c){
    float w = ekf->params->robot_width;
    if (fabsf(c->r > c->l) > 0.00001){
        float rml = c->r - c->l;
        float rml2 = rml * rml;
        float theta_ = ekf->theta + rml/w;
        float rpl = c->r + c->l;
        
        printf("-->: %f %f %f %f\r\n", rml, rml2, theta_, rpl);

        ekf->V[0][0] = w*c->r*(sin(theta_)-sin(ekf->theta))/rml2 - rpl*cos(theta_)/(2*rml);
        ekf->V[1][0] = w*c->r*(-cos(theta_)+cos(ekf->theta))/rml2 - rpl*sin(theta_)/(2*rml);
        ekf->V[2][0] = -1.0f/w;

        ekf->V[0][1] = (-w*c->l)/rml2*(sin(theta_)-sin(ekf->theta)) + rpl/(2*rml)*cos(theta_);
        ekf->V[1][1] = (-w*c->l)/rml2*(-cos(theta_)+cos(ekf->theta)) + rpl/(2*rml)*sin(theta_);
        ekf->V[2][1] = 1.0f/w;
    }
    else{
        ekf->V[0][0] = 0.5f*(cos(ekf->theta) + c->l/w*sin(ekf->theta));
        ekf->V[1][0] = 0.5f*(sin(ekf->theta) - c->l/w*cos(ekf->theta));
        ekf->V[2][0] = -1.0f/w;

        ekf->V[0][1] = 0.5f*(-c->l/w*sin(ekf->theta) + cos(ekf->theta));
        ekf->V[1][1] = 0.5f*(c->l/w*cos(ekf->theta) + sin(ekf->theta));
        ekf->V[2][1] = 1.0f/w;
    }
    printf("V: %f %f %f %f %f %f\r\n", ekf->V[0][0], ekf->V[0][1], ekf->V[1][0], ekf->V[1][1], ekf->V[2][0], ekf->V[2][1]);
}

void ekf_h(ekf_t *ekf, ekf_measurement_t *m){
    float dx = m->rx - (ekf->x + ekf->params->tag_polar_r* cos(ekf->theta + ekf->params->tag_polar_theta));
    float dy = m->ry - (ekf->y + ekf->params->tag_polar_r * sin(ekf->theta + ekf->params->tag_polar_theta));
    
    ekf->h[0] = sqrt(dx * dx + dy * dy);
    ekf->h[1] = ekf->theta;
}

void ekf_dh_dstate(ekf_t *ekf, ekf_measurement_t *m){
    float dx = m->rx - (ekf->x + ekf->params->tag_polar_r* cos(ekf->theta + ekf->params->tag_polar_theta));
    float dy = m->ry - (ekf->y + ekf->params->tag_polar_r * sin(ekf->theta + ekf->params->tag_polar_theta));
    float r = sqrt(dx * dx + dy * dy);

    ekf->H[0][0] = -dx / r;
    ekf->H[0][1] = -dy / r;
    ekf->H[0][2] = (dx * sin(ekf->theta + ekf->params->tag_polar_theta) - dy * cos(ekf->theta + ekf->params->tag_polar_theta)) * ekf->params->tag_polar_r / r;
    ekf->H[1][0] = ekf->H[1][1] = 0;
    ekf->H[1][2] = 1;
}

File: software/ekf/ekf_real/ekf.py
from math import *
from numpy import array, dot, eye, zeros, linalg, size, diag

class ExtendedKalmanFilter:

    def __init__(self, state, covariance):
        self.state = state
        self.covariance = covariance

    def set_control_params(self, robot_width, control_motion_factor, control_turn_factor):
        self.robot_width = robot_width
        self.control_motion_factor = control_motion_factor
        self.control_turn_factor = control_turn_factor

    def set_measurement_params(self, scanner_displacement, measurement_distance_stddev, measurement_angle_stddev):
        self.scanner_displacement = scanner_displacement
        self.measurement_distance_stddev = measurement_distance_stddev
        self.measurement_angle_stddev = measurement_angle_stddev

    @staticmethod
    def g(state, control, w):
        x, y, theta = state
        l, r = control
        if r != l:
            alpha = (r - l) / w
            rad = l/alpha
            g1 = x + (rad + w/2.)*(sin(theta+alpha) - sin(theta))
            g2 = y + (rad + w/2.)*(-cos(theta+alpha) + cos(theta))
            g3 = (theta + alpha + pi) % (2*pi) - pi
        else:
            g1 = x + l * cos(theta)
            g2 = y + l * sin(theta)
            g3 = theta

        return array([g1, g2, g3])

    @staticmethod
    def dg_dstate(state, control, w):
        theta = state[2]
        l, r = control
        if r != l:
            alpha = (r-l)/w
            theta_ = theta + alpha
            rpw2 = l/alpha + w/2.0
            m = array([[1.0, 0.0, rpw2*(cos(theta_) - cos(theta))],
                       [0.0, 1.0, rpw2*(sin(theta_) - sin(theta))],
                       [0.0, 0.0, 1.0]])
        else:
            m = array([[1.0, 0.0, -l*sin(theta)],
                       [0.0, 1.0,  l*cos(theta)],
                       [0.0, 0.0,  1.0]])
        return m

    @staticmethod
    def dg_dcontrol(state, control, w):
        theta = state[2]
        l, r = tuple(control)
        if r != l:
            rml = r - l
            rml2 = rml * rml
            theta_ = theta + rml/w
            dg1dl = w*r/rml2*(sin(theta_)-sin(theta))  - (r+l)/(2*rml)*cos(theta_)
            dg2dl = w*r/rml2*(-cos(theta_)+cos(theta)) - (r+l)/(2*rml)*sin(theta_)
            dg1dr = (-w*l)/rml2*(sin(theta_)-sin(theta)) + (r+l)/(2*rml)*cos(theta_)
            dg2dr = (-w*l)/rml2*(-cos(theta_)+cos(theta)) + (r+l)/(2*rml)*sin(theta_)
            
        else:
            dg1dl = 0.5*(cos(theta) + l/w*sin(theta))
            dg2dl = 0.5*(sin(theta) - l/w*cos(theta))
            dg1dr = 0.5*(-l/w*sin(theta) + cos(theta))
            dg2dr = 0.5*(l/w*cos(theta) + sin(theta))

        dg3dl = -1.0/w
        dg3dr = 1.0/w
        m = array([[dg1dl, dg1dr], [dg2dl, dg2dr], [dg3dl, dg3dr]])

        return m

    def predict(self, control):
        left, right = control
        sigma_l_square = (self.control_motion_factor*left)**2 + (self.control_turn_factor*(left-right))**2
        sigma_r_square = (self.control_motion_factor*right)**2 + (self.control_turn_factor*(right-left))**2
        self.control_covariance = diag([sigma_l_square, sigma_r_square])

        G = self.dg_dstate(self.state, control, self.robot_width)
        V = self.dg_dcontrol(self.state, control, self.robot_width)

        self.state = self.g(self.state, control, self.robot_width)
        self.covariance = dot(dot(G,self.covariance), G.T) + dot(dot(V,self.control_covariance),V.T)

    @staticmethod
    def h(state, landmark, scanner_displacement):
        dx = landmark[0] - (state[0] + scanner_displacement * cos(state[2]))
        dy = landmark[1] - (state[1] + scanner_displacement * sin(state[2]))
        r = sqrt(dx * dx + dy * dy)
        alpha = state[2]

        return array([r, alpha])

    @staticmethod
    def dh_dstate(state, landmark, scanner_displacement):
        theta = state[2]
        cos_t, sin_t = cos(theta), sin(theta)
        dx = landmark[0] - (state[0] + scanner_displacement * cos_t)
        dy = landmark[1] - (state[1] + scanner_displacement * sin_t)
        q = dx * dx + dy * dy
        sqrt_q = sqrt(q)
        dr_dx = -dx / sqrt_q
        dr_dy = -dy / sqrt_q
        dr_dtheta = (dx * sin_t - dy * cos_t) * scanner_displacement / sqrt_q

        return array([[dr_dx, dr_dy, dr_dtheta], [0,0,1]])

    def correct(self, measurement, landmark):
        H = self.dh_dstate(self.state, landmark, self.scanner_displacement)
        Q = array([[self.measurement_distance_stddev**2,0],[0,self.measurement_angle_stddev**2]])
        
        innovation = array(measurement) - self.h(self.state, landmark, self.scanner_displacement)
        innovation[1] = (innovation[1] + pi) % (2*pi) - pi

        # print(self.state)
        # print(landmark)
        # print(measurement)
        # print(self.h(self.state, landmark, self.scanner_displacement))
        # print(innovation)

        K = dot(self.covariance,dot(H.T, linalg.inv(dot(H, dot(self.covariance,H.T))+Q)))
        self.state = self.state + dot(K, innovation)
        self.covariance = dot(eye(3) - dot(K,H), self.covariance)

        # print(self.state)
        # print("-------")

    def get_state(self):
        return self.state[0:3]

    def get_state_covariance(self):
        return self.covariance[0:3, 0:3]

    @staticmethod
    def get_error_ellipse(covariance):
        """Return the position covariance (which is the upper 2x2 submatrix)
           as a triple: (main_axis_angle, stddev_1, stddev_2), where
           main_axis_angle is the angle (pointing direction) of the main axis,
           along which the standard deviation is stddev_1, and stddev_2 is the
           standard deviation along the other (orthogonal) axis."""
        eigenvals, eigenvects = linalg.eig(covariance[0:2,0:2])
        angle = atan2(eigenvects[1,0], eigenvects[0,0])
        return [angle, sqrt(eigenvals[0]), sqrt(eigenvals[1])]
